### PR TITLE
fix: `eventsRef` should trigger the lastest version of config callbacks

### PR DESCRIPTION
**Pull request by lintuming**
# fix

Resolve: #508 #515 
 
Now the `eventsRef` read the config via the `ref`,which always capture the lastest version of the config callbacks.

This fix the problem of the stale closure related to the config callbacks.

### CODE REVIEW

**Comment by lintuming** [comment]
cc @sergiodxa

**Comment by lintuming** [comment]
Also, i've noticed some code paths that using the `config` may have some similar stale closure problem.
Should we consider the scenario the config passed to the `useSWR` hook may change along the component lifecycle?

File: src/use-swr.ts [hunk 1]
```diff
@@ -192,6 +192,11 @@ function useSWR<Data = any, Error = any>(
     config
   )
 
+  const lastestConfig = useRef(config)
+  useIsomorphicLayoutEffect(() => {
```
Comment (huozhi) [inline src/use-swr.ts:196]: I'm not sure if we need to have this useEffect here, could it just be updated in each rendering directly?

```js
const configRef = useRef(config)
configRef.current = config
```

Comment (shuding) [reply src/use-swr.ts]: @huozhi I feel that we still need to wrap it with `useEffect` (`useIsomorphicLayoutEffect` here). Updating refs directly in render can be dangerous in Concurrent Mode (not 100% about this specific usage here). 

I'll approve this for now, but I suggest adding a comment to clarify :+1:

Comment (huozhi) [reply src/use-swr.ts]: sure, good to know there's side effect on concurrent mode! 🙏 sorry for my unreasonable comment above

Comment (lintuming) [reply src/use-swr.ts]: @shuding thanks to point this out,do i need to change this back?

Comment (huozhi) [reply src/use-swr.ts]: yep, I think you can just simply revert the last commit. the rest looks good 👍

**Comment by nicholaschiang** [comment]
@shuding @huozhi @lintuming could we merge this?

**Pull request merged**